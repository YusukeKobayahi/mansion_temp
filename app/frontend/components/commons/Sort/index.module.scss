@import "~/styles/variables";
@import "~/styles/utils";

.container {
  display: flex;
  align-items: center;
  margin-left: 30px;

  @include screenMd {
    width: 100%;
    justify-content: flex-end;
    margin: 12px 0 0;
  }

  p {
    color: #746744;
    font-size: 16px;
    font-weight: bold;
  }

  .sortBox {
    width: 160px;
    height: 50px;
    position: relative;
    margin-left: 16px;

    @include screenMd {
      width: 180px;
    }

    select {
      width: 100%;
      height: 100%;
      appearance: none;
      color: #746744;
      font-size: 14px;
      font-weight: bold;
      background-color: #ffffff;
      padding: 0 8px;
      background-image: url("/ms-library/images/select.png");
      background-size: 12px 10px;
      background-position: right 5px center;
      background-repeat: no-repeat;
      border: 1px solid #767676;

      @include screenMd {
        font-size: 16px;
      }
    }
  }
}
