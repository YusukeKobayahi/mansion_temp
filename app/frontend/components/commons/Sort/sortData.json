{
  "orud": [
    {
      "name": "新着・更新順",
      "value": "orud&ASC"
    }
  ],
  "orpr": [
    {
      "name": "価格が安い順",
      "value": "orpr&ASC"
    },
    {
      "name": "価格が高い順",
      "value": "orpr&DESC"
    }
  ],
  "orar": [
    {
      "name": "占有面積が狭い順",
      "value": "orar&ASC"
    },
    {
      "name": "占有面積が広い順",
      "value": "orar&DESC"
    }
  ],
  "orck": [
    {
      "name": "築年月が新しい順",
      "value": "orck&ASC"
    },
    {
      "name": "築年月が古い順",
      "value": "orck&DESC"
    }
  ],
  "ormd": [
    {
      "name": "間取りが少ない順",
      "value": "ormd&ASC"
    },
    {
      "name": "間取りが多い順",
      "value": "ormd&DESC"
    }
  ]
}
