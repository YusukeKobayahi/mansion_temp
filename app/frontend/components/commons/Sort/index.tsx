import Router from "next/router";
import includes from "lodash/includes";
import styles from "~/components/commons/Sort/index.module.scss";
import { QueryVariables } from "~/lib/types";
import { searchOptions, SortQueryKeys, ModelQueryKeys } from "~/lib/property";
import map from "lodash/map";
import omitBy from "lodash/omitBy";
import intersection from "lodash/intersection";
import sortData from "~/components/commons/Sort/sortData.json";

export interface SortProps {
  pathname: string;
  searchCondition: QueryVariables;
}

const Sort: React.FC<SortProps> = ({ pathname, searchCondition }) => {
  const currentSortKeys = intersection(
    SortQueryKeys,
    Object.keys(searchCondition)
  );
  const defaultValue = (() => {
    if (!currentSortKeys.length) sortData.orud[0].value;
    const key = currentSortKeys[0] as keyof QueryVariables;
    return `${key}&${searchCondition[key]}`;
  })();

  const query = omitBy(
    { ...searchCondition },
    (value, key) =>
      includes(searchOptions, key) ||
      (includes(ModelQueryKeys, key) && String(value).split(",").length === 1)
  );

  const setQuery = (key: string, value: string) => {
    query[key] = value;
    Router.push({
      pathname,
      query,
    });
  };

  const sortItem = map(sortData, (value) => {
    return value.map((data, i) => (
      <option key={i} value={data.value}>
        {data.name}
      </option>
    ));
  });
  // onChangeで並び替えの即時検索を行いためESLintを無効に
  /* eslint-disable */ return (
    <section className={styles.container}>
      <p>並べ替え</p>
      <div className={styles.sortBox}>
        <select
          name="sort"
          onChange={(e) => {
            setQuery(
              e.target.value.split("&")[0],
              e.target.value.split("&")[1]
            );
          }}
          value={defaultValue}
        >
          {sortItem}
        </select>
      </div>
    </section>
  );
};

export default Sort;
