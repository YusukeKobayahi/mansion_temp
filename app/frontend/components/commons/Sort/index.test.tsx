import { render } from "@testing-library/react";
import Sort from "~/components/commons/Sort";

describe("components/commons/Sort", () => {
  it("正常に表示される", () => {
    const pathname = "/cities/13101/";
    const searchCondition = {
      gft: 5,
    };
    const { asFragment } = render(
      <Sort pathname={pathname} searchCondition={searchCondition} />
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
