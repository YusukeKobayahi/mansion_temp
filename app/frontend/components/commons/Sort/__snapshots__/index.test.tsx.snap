// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/commons/Sort 正常に表示される 1`] = `
<DocumentFragment>
  <section
    class="container"
  >
    <p>
      並べ替え
    </p>
    <div
      class="sortBox"
    >
      <select
        name="sort"
      >
        <option
          value="orud&ASC"
        >
          新着・更新順
        </option>
        <option
          value="orpr&ASC"
        >
          価格が安い順
        </option>
        <option
          value="orpr&DESC"
        >
          価格が高い順
        </option>
        <option
          value="orar&ASC"
        >
          占有面積が狭い順
        </option>
        <option
          value="orar&DESC"
        >
          占有面積が広い順
        </option>
        <option
          value="orck&ASC"
        >
          築年月が新しい順
        </option>
        <option
          value="orck&DESC"
        >
          築年月が古い順
        </option>
        <option
          value="ormd&ASC"
        >
          間取りが少ない順
        </option>
        <option
          value="ormd&DESC"
        >
          間取りが多い順
        </option>
      </select>
    </div>
  </section>
</DocumentFragment>
`;
